tests/phpunit/tests/admin/wpAutomaticUpdater.php

```diff
@@ -0,0 +1,572 @@
+<?php
+
+/**
+ * @group upgrade
+ *
+ * @covers WP_Automatic_Updater
+ */
+class Tests_Admin_WpAutomaticUpdater extends WP_UnitTestCase {
+	/**
+	 * An instance of WP_Automatic_Updater.
+	 *
+	 * @var WP_Automatic_Updater
+	 */
+	private static $updater;
+
+	/**
+	 * WP_Automatic_Updater::send_plugin_theme_email
+	 * made accessible.
+	 *
+	 * @var ReflectionMethod
+	 */
+	private static $send_plugin_theme_email;
+
+	/**
+	 * Sets up shared fixtures.
+	 */
+	public static function set_up_before_class() {
+		require_once ABSPATH . 'wp-admin/includes/class-wp-automatic-updater.php';
+		self::$updater = new WP_Automatic_Updater();
+
+		self::$send_plugin_theme_email = new ReflectionMethod( self::$updater, 'send_plugin_theme_email' );
+		self::$send_plugin_theme_email->setAccessible( true );
+
+		add_filter( 'pre_wp_mail', '__return_false' );
+	}
+
+	/**
+	 * Tests that `WP_Automatic_Updater::send_plugin_theme_email()` appends
+	 * plugin URLs.
+	 *
+	 * @ticket 53049
+	 *
+	 * @covers WP_Automatic_Updater::send_plugin_theme_email
+	 *
+	 * @dataProvider data_send_plugin_theme_email_should_append_plugin_urls
+	 *
+	 * @param string[] $urls       The URL(s) to search for. Must not be empty.
+	 * @param object[] $successful An array of successful plugin update objects.
+	 * @param object[] $failed     An array of failed plugin update objects.
+	 */
+	public function test_send_plugin_theme_email_should_append_plugin_urls( $urls, $successful, $failed ) {
+		add_filter(
+			'wp_mail',
+			function( $args ) use ( $urls ) {
+				foreach ( $urls as $url ) {
+					$this->assertStringContainsString(
+						$url,
+						$args['message'],
+						'The email message should contain ' . $url
+					);
+				}
+			}
+		);
+
+		$has_successful = ! empty( $successful );
+		$has_failed     = ! empty( $failed );
+
+		if ( ! $has_successful && ! $has_failed ) {
+			$this->markTestSkipped( 'This test requires at least one successful or failed plugin update object.' );
+		}
+
+		$type = $has_successful && $has_failed ? 'mixed' : ( ! $has_failed ? 'success' : 'fail' );
```

Comment:
This isn't strictly necessary and could be replaced with a `$type` argument in _each dataset_.

I included this in the tests to:

- let contributors could focus on testing the update objects.
- reduce the size of the datasets.
- reduce the potential for human error by failing to note the correct `$type` when copying/pasting datasets.